### PR TITLE
feat: support theming solara

### PR DESCRIPTION
**Pull request by iisakkirotko**

### CODE REVIEW

**Comment by iisakkirotko** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/widgetti/solara/494?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #471** <a href="https://app.graphite.dev/github/pr/widgetti/solara/471?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #455** <a href="https://app.graphite.dev/github/pr/widgetti/solara/455?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #461** <a href="https://app.graphite.dev/github/pr/widgetti/solara/461?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #494** <a href="https://app.graphite.dev/github/pr/widgetti/solara/494?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
          * **PR #495** <a href="https://app.graphite.dev/github/pr/widgetti/solara/495?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #473** <a href="https://app.graphite.dev/github/pr/widgetti/solara/473?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #474** <a href="https://app.graphite.dev/github/pr/widgetti/solara/474?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #452** <a href="https://app.graphite.dev/github/pr/widgetti/solara/452?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/widgetti/solara/494?utm_source=stack-comment).

File: solara/lab/components/theming.vue [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            :color="theme_effective ? 'primary' : null">
+            {{ this.clicks === 1 ? this.on_icon : this.clicks === 2 ? this.off_icon : this.auto_icon }}
+        </v-icon>
+    </v-btn>
+</template>
+<script>
+module.exports = {
+    mounted() {
+        if (theme.variant_user_selectable) {
+            if (localStorage.getItem('theme.variant')) {
```
Comment (maartenbreddels) [inline solara/lab/components/theming.vue:16]:
```suggestion
            if (localStorage.getItem(':solara:theme.variant')) {
```
Shall we prefix it with something like this? Other prefix ideas are welcome :)
This reduces the chance of future name collisions with user chosen names

File: solara/server/app.py [hunk 2]
```diff
@@ -350,12 +350,17 @@ def on_msg(msg):
...
+            path = args.get("path", "")
+            app_name = args.get("appName") or "__default__"
             app = apps[app_name]
             context = kernel_context.get_current_context()
+            dark = args.get("dark", False)
             import ipyvuetify
 
+            from solara.lab import theme
+
+            theme.dark_effective = dark
```
Comment (maartenbreddels) [inline solara/server/app.py:363]:
```suggestion
            # While this usually gets set from the frontend, in solara (server) we want to know this directly at the first
            # render. Also, using the same trait allows us to write code which works on all widgets platforms, instead
            # or using something different when running under solara server
            theme.dark_effective = dark
```

File: solara/website/pages/api/theming.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+"""
+# Themes in Solara
+
+Theming is provided to Solara through the [`ipyvuetify`](https://solara.dev/docs/understanding/ipyvuetify) package.
+Two themes are provided by default: light and dark.
```
[the remark this answers is not in the document]
Comment (maartenbreddels) [reply solara/website/pages/api/theming.py:5]: I suggest we change the wording to say, 'theme; and 'dark', and in the settings it's `solara_theme_dark`. I think nobody knows what 'variant' means without spending brain cycles.
Also, does vuetify 3 change this?

**Review by maartenbreddels — COMMENTED** [review]
I think this is going to be super cool!
Mostly worried about the naming, and maybe the theming.vue can be cleaned up a bit by using computed properties instead of using the clicks directly (difficult to follow).
(e.g. using a .is_auto() { return ...} )

File: solara/website/pages/api/theming.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (iisakkirotko) [reply solara/website/pages/api/theming.py:5]: `Vuetify supports both light and dark variants of the Material Design spec` Is what vuetify says in vuetify 2. In vuetify 3 it seems like they've gotten rid of the word variant, there the page says that they support two themes, light and dark. I think it's probably best for us to adopt the language from vuetify 3.

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Looking good.
I think we should remove totally the `variant_user_selectable` and change
```
class ThemeSettings(BaseSettings):
    variant: ThemeVariant = ThemeVariant.light
```

to
```
class ThemeSettings(BaseSettings):
    dark: ThemeVariant = ThemeVariant.light
```

To make it backward compatible, we should check if `SOLARA_THEME_VARIANT` is set, and make theme.dark reflect that.

Also, think about how this plays with the cmd line arg `"--theme-variant`. Not sure how we should expose it with the rename, `--dark=dark` seems a bit odd.
This could be a reason to keep the name variant, let me know what you think

**Pull request merged**